File: finstack-py/tests/golden/runners/calibration_common.py
"""Executable helpers for calibration golden fixtures."""

from __future__ import annotations

from datetime import date
import math
from typing import Any

from finstack.core.market_data import DiscountCurve, ForwardCurve, HazardCurve, InflationCurve

from tests.golden.runners import validate_source_validation_fixture
from tests.golden.schema import GoldenFixture


def run_curve_fixture(fixture: GoldenFixture) -> dict[str, float]:
    validate_source_validation_fixture("curve calibration runner", fixture)
    inputs = fixture.inputs
    discounts = {curve["id"]: curve for curve in inputs.get("discount", [])}
    forwards = {curve["id"]: curve for curve in inputs.get("forward", [])}
    inflations = {curve["id"]: curve for curve in inputs.get("inflation", [])}
    discount_curves = {curve_id: _build_discount_curve(curve) for curve_id, curve in discounts.items()}
    forward_curves = {curve_id: _build_forward_curve(curve) for curve_id, curve in forwards.items()}
    inflation_curves = {curve_id: _build_inflation_curve(curve) for curve_id, curve in inflations.items()}
    actuals: dict[str, float] = {}
    for probe in inputs.get("probes", []):
        curve_id = probe["curve"]
        tenor = float(probe["tenor"])
        kind = probe["kind"]
        if kind == "discount_factor":
            actuals[probe["output"]] = float(discount_curves[curve_id].df(tenor))
        elif kind == "zero_rate":
            actuals[probe["output"]] = float(discount_curves[curve_id].zero(tenor))
        elif kind == "forward_rate":
            actuals[probe["output"]] = float(forward_curves[curve_id].rate(tenor))
        elif kind == "cpi":
            actuals[probe["output"]] = float(inflation_curves[curve_id].cpi(tenor))
        elif kind == "inflation_zero_rate":
            curve = inflations[curve_id]
            cpi = inflation_curves[curve_id].cpi(tenor)
            actuals[probe["output"]] = math.log(cpi / float(curve["base_cpi"])) / max(tenor, 1e-12)
        else:
            msg = f"unsupported curve probe kind {kind!r}"
            raise ValueError(msg)
    return _reject_non_executable_calibration("curve calibration runner", fixture)


def run_hazard_fixture(fixture: GoldenFixture) -> dict[str, float]:
    validate_source_validation_fixture("hazard calibration runner", fixture)
    inputs = fixture.inputs
    hazards = {curve["id"]: curve for curve in inputs.get("hazard", [])}
    hazard_curves = {curve_id: _build_hazard_curve(curve) for curve_id, curve in hazards.items()}
    actuals: dict[str, float] = {}
    for probe in inputs.get("probes", []):
        curve = hazard_curves[probe["curve"]]
        tenor = float(probe["tenor"])
        if probe["kind"] == "hazard_rate":
            actuals[probe["output"]] = float(curve.hazard_rate(tenor))
        elif probe["kind"] == "survival_probability":
            actuals[probe["output"]] = float(curve.survival(tenor))
        else:
            msg = f"unsupported hazard probe kind {probe['kind']!r}"
            raise ValueError(msg)
    return _reject_non_executable_calibration("hazard calibration runner", fixture)


def run_vol_smile_fixture(fixture: GoldenFixture) -> dict[str, float]:
    validate_source_validation_fixture("vol smile calibration runner", fixture)
    actuals: dict[str, float] = {}
    for smile in fixture.inputs["smiles"]:
        prefix = f"{smile['id']}::{smile['expiry']}"
        atm = float(smile["atm_vol"])
        put25 = float(smile["wing_25d_put_vol"])
        call25 = float(smile["wing_25d_call_vol"])
        actuals[f"atm_vol::{prefix}"] = atm
        actuals[f"wing_25d_put_vol::{prefix}"] = put25
        actuals[f"wing_25d_call_vol::{prefix}"] = call25
        actuals[f"risk_reversal_25d::{prefix}"] = call25 - put25
        actuals[f"butterfly_25d::{prefix}"] = 0.5 * (call25 + put25) - atm
        if "wing_10d_put_vol" in smile and "wing_10d_call_vol" in smile:
            put10 = float(smile["wing_10d_put_vol"])
            call10 = float(smile["wing_10d_call_vol"])
            actuals[f"risk_reversal_10d::{prefix}"] = call10 - put10
            actuals[f"butterfly_10d::{prefix}"] = 0.5 * (call10 + put10) - atm
    return _reject_non_executable_calibration("vol smile calibration runner", fixture)


def run_sabr_cube_fixture(fixture: GoldenFixture) -> dict[str, float]:
    validate_source_validation_fixture("SABR calibration runner", fixture)
    parameters = {key: float(value) for key, value in fixture.inputs["parameters"].items()}
    if not parameters:
        msg = "SABR source validation requires committed parameters"
        raise ValueError(msg)
    return _reject_non_executable_calibration("SABR calibration runner", fixture)


def _reject_non_executable_calibration(runner: str, fixture: GoldenFixture) -> dict[str, float]:
    if any(key in fixture.inputs for key in ("quotes", "market_quotes", "calibration_quotes")):
        quote_hint = (
            " quote inputs are present, but this golden runner has not wired them to a product calibration engine yet."
        )
    else:
        quote_hint = (
            " fixture contains final curves/parameters but no quote/source instruments from which to calibrate."
        )
    msg = (
        f"{runner} requires executable quote inputs and must compute RMSE from calibrated outputs;{quote_hint} "
        "Reclassify as source_validation or add a real calibration contract."
    )
    raise ValueError(msg)


def _build_discount_curve(spec: dict[str, Any]) -> DiscountCurve:
    return DiscountCurve(
        id=spec["id"],
        base_date=date.fromisoformat(spec["base_date"]),
        knots=_knots(spec["knots"]),
        interp=spec.get("interp", "linear"),
        day_count=spec.get("day_count"),
    )


def _build_forward_curve(spec: dict[str, Any]) -> ForwardCurve:
    return ForwardCurve(
        id=spec["id"],
        tenor=float(spec["tenor"]),
        base_date=date.fromisoformat(spec["base_date"]),
        knots=_knots(spec["knots"]),
        interp=spec.get("interp", "linear"),
        day_count=spec.get("day_count", "act_360"),
    )


def _build_inflation_curve(spec: dict[str, Any]) -> InflationCurve:
    return InflationCurve(
        id=spec["id"],
        base_date=date.fromisoformat(spec["base_date"]),
        base_cpi=float(spec["base_cpi"]),
        knots=_knots(spec["knots"]),
        day_count=spec.get("day_count", "act_365f"),
        indexation_lag_months=int(spec.get("indexation_lag_months", 3)),
        interp=spec.get("interp", "log_linear"),
    )


def _build_hazard_curve(spec: dict[str, Any]) -> HazardCurve:
    return HazardCurve(
        id=spec["id"],
        base_date=date.fromisoformat(spec["base_date"]),
        knots=_knots(spec["knots"]),
        recovery_rate=spec.get("recovery_rate"),
        day_count=spec.get("day_count", "act_365f"),
    )


def _knots(raw_knots: list[list[float]]) -> list[tuple[float, float]]:
    return [(float(t), float(value)) for t, value in raw_knots]

File: finstack/valuations/tests/golden/runners/calibration_common.rs
//! Shared executable calibration fixture helpers.

use crate::golden::schema::GoldenFixture;
use finstack_core::dates::DayCount;
use finstack_core::market_data::term_structures::{
    DiscountCurve, ForwardCurve, HazardCurve, InflationCurve,
};
use finstack_core::math::interp::InterpStyle;
use finstack_valuations::pricer::parse_as_of_date;
use serde::Deserialize;
use std::collections::BTreeMap;
use std::str::FromStr;

#[derive(Debug, Deserialize)]
pub(crate) struct CurveCalibrationInputs {
    #[serde(default)]
    discount: Vec<DiscountCurveSpec>,
    #[serde(default)]
    forward: Vec<ForwardCurveSpec>,
    #[serde(default)]
    inflation: Vec<InflationCurveSpec>,
    #[serde(default)]
    probes: Vec<ProbeSpec>,
    #[serde(default)]
    calibration_rmse: Option<f64>,
}

#[derive(Debug, Deserialize)]
pub(crate) struct HazardCalibrationInputs {
    hazard: Vec<HazardCurveSpec>,
    #[serde(default)]
    probes: Vec<ProbeSpec>,
    #[serde(default)]
    calibration_rmse: Option<f64>,
}

#[derive(Debug, Deserialize)]
pub(crate) struct VolSmileInputs {
    smiles: Vec<VolSmileSpec>,
    #[serde(default)]
    repriced_rmse: Option<f64>,
}

#[derive(Debug, Deserialize)]
pub(crate) struct SabrCubeInputs {
    parameters: BTreeMap<String, f64>,
    #[serde(default)]
    calibration_rmse: Option<f64>,
}

#[derive(Debug, Deserialize)]
struct DiscountCurveSpec {
    id: String,
    base_date: String,
    day_count: Option<String>,
    interp: Option<String>,
    knots: Vec<[f64; 2]>,
}

#[derive(Debug, Deserialize)]
struct ForwardCurveSpec {
    id: String,
    tenor: f64,
    base_date: String,
    day_count: Option<String>,
    interp: Option<String>,
    knots: Vec<[f64; 2]>,
}

#[derive(Debug, Deserialize)]
struct InflationCurveSpec {
    id: String,
    base_date: String,
    base_cpi: f64,
    day_count: Option<String>,
    interp: Option<String>,
    indexation_lag_months: Option<u32>,
    knots: Vec<[f64; 2]>,
}

#[derive(Debug, Deserialize)]
struct HazardCurveSpec {
    id: String,
    base_date: String,
    recovery_rate: Option<f64>,
    day_count: Option<String>,
    knots: Vec<[f64; 2]>,
}

#[derive(Debug, Deserialize)]
struct ProbeSpec {
    output: String,
    curve: String,
    kind: String,
    tenor: f64,
}

#[derive(Debug, Deserialize)]
struct VolSmileSpec {
    id: String,
    expiry: String,
    atm_vol: f64,
    wing_25d_put_vol: f64,
    wing_25d_call_vol: f64,
    wing_10d_put_vol: Option<f64>,
    wing_10d_call_vol: Option<f64>,
}

pub(crate) fn run_curve_fixture(fixture: &GoldenFixture) -> Result<BTreeMap<String, f64>, String> {
    let inputs: CurveCalibrationInputs = serde_json::from_value(fixture.inputs.clone())
        .map_err(|err| format!("parse curve calibration inputs: {err}"))?;
    let mut discount_curves = BTreeMap::new();
    for curve in &inputs.discount {
        discount_curves.insert(curve.id.clone(), build_discount_curve(curve)?);
    }
    let mut forward_curves = BTreeMap::new();
    for curve in &inputs.forward {
        forward_curves.insert(curve.id.clone(), build_forward_curve(curve)?);
    }
    let mut inflation_curves = BTreeMap::new();
    let mut inflation_base_cpi = BTreeMap::new();
    for curve in &inputs.inflation {
        inflation_base_cpi.insert(curve.id.clone(), curve.base_cpi);
        inflation_curves.insert(curve.id.clone(), build_inflation_curve(curve)?);
    }

    let mut actuals = BTreeMap::new();
    for probe in &inputs.probes {
        let value = match probe.kind.as_str() {
            "discount_factor" => discount_curves
                .get(&probe.curve)
                .ok_or_else(|| format!("missing discount curve '{}'", probe.curve))?
                .df(probe.tenor),
            "zero_rate" => discount_curves
                .get(&probe.curve)
                .ok_or_else(|| format!("missing discount curve '{}'", probe.curve))?
                .zero(probe.tenor),
            "forward_rate" => forward_curves
                .get(&probe.curve)
                .ok_or_else(|| format!("missing forward curve '{}'", probe.curve))?
                .rate(probe.tenor),
            "cpi" => inflation_curves
                .get(&probe.curve)
                .ok_or_else(|| format!("missing inflation curve '{}'", probe.curve))?
                .cpi(probe.tenor),
            "inflation_zero_rate" => {
                let curve = inflation_curves
                    .get(&probe.curve)
                    .ok_or_else(|| format!("missing inflation curve '{}'", probe.curve))?;
                let base_cpi = inflation_base_cpi
                    .get(&probe.curve)
                    .ok_or_else(|| format!("missing inflation base CPI '{}'", probe.curve))?;
                (curve.cpi(probe.tenor) / base_cpi).ln() / probe.tenor.max(1e-12)
            }
            other => return Err(format!("unsupported curve probe kind '{other}'")),
        };
        actuals.insert(probe.output.clone(), value);
    }
    if let Some(rmse) = inputs.calibration_rmse {
        actuals.insert("calibration_rmse".to_string(), rmse);
    }
    Ok(actuals)
}

pub(crate) fn run_hazard_fixture(fixture: &GoldenFixture) -> Result<BTreeMap<String, f64>, String> {
    let inputs: HazardCalibrationInputs = serde_json::from_value(fixture.inputs.clone())
        .map_err(|err| format!("parse hazard calibration inputs: {err}"))?;
    let mut hazard_curves = BTreeMap::new();
    for curve in &inputs.hazard {
        hazard_curves.insert(curve.id.clone(), build_hazard_curve(curve)?);
    }

    let mut actuals = BTreeMap::new();
    for probe in &inputs.probes {
        let curve = hazard_curves
            .get(&probe.curve)
            .ok_or_else(|| format!("missing hazard curve '{}'", probe.curve))?;
        let value = match probe.kind.as_str() {
            "hazard_rate" => curve.hazard_rate(probe.tenor),
            "survival_probability" => curve.sp(probe.tenor),
            other => return Err(format!("unsupported hazard probe kind '{other}'")),
        };
        actuals.insert(probe.output.clone(), value);
    }
    if let Some(rmse) = inputs.calibration_rmse {
        actuals.insert("calibration_rmse".to_string(), rmse);
    }
    Ok(actuals)
}

pub(crate) fn run_vol_smile_fixture(
    fixture: &GoldenFixture,
) -> Result<BTreeMap<String, f64>, String> {
    let inputs: VolSmileInputs = serde_json::from_value(fixture.inputs.clone())
        .map_err(|err| format!("parse vol smile inputs: {err}"))?;
    let mut actuals = BTreeMap::new();
    for smile in inputs.smiles {
        let prefix = format!("{}::{}", smile.id, smile.expiry);
        actuals.insert(format!("atm_vol::{prefix}"), smile.atm_vol);
        actuals.insert(
            format!("wing_25d_put_vol::{prefix}"),
            smile.wing_25d_put_vol,
        );
        actuals.insert(
            format!("wing_25d_call_vol::{prefix}"),
            smile.wing_25d_call_vol,
        );
        actuals.insert(
            format!("risk_reversal_25d::{prefix}"),
            smile.wing_25d_call_vol - smile.wing_25d_put_vol,
        );
        actuals.insert(
            format!("butterfly_25d::{prefix}"),
            0.5 * (smile.wing_25d_call_vol + smile.wing_25d_put_vol) - smile.atm_vol,
        );
        if let (Some(put), Some(call)) = (smile.wing_10d_put_vol, smile.wing_10d_call_vol) {
            actuals.insert(format!("risk_reversal_10d::{prefix}"), call - put);
            actuals.insert(
                format!("butterfly_10d::{prefix}"),
                0.5 * (call + put) - smile.atm_vol,
            );
        }
    }
    if let Some(rmse) = inputs.repriced_rmse {
        actuals.insert("repriced_rmse".to_string(), rmse);
    }
    Ok(actuals)
}

pub(crate) fn run_sabr_cube_fixture(
    fixture: &GoldenFixture,
) -> Result<BTreeMap<String, f64>, String> {
    let inputs: SabrCubeInputs = serde_json::from_value(fixture.inputs.clone())
        .map_err(|err| format!("parse SABR cube inputs: {err}"))?;
    let mut actuals = inputs.parameters;
    if let Some(rmse) = inputs.calibration_rmse {
        actuals.insert("calibration_rmse".to_string(), rmse);
    }
    Ok(actuals)
}

fn build_discount_curve(spec: &DiscountCurveSpec) -> Result<DiscountCurve, String> {
    let mut builder = DiscountCurve::builder(spec.id.as_str())
        .base_date(parse_as_of_date(&spec.base_date).map_err(|err| err.to_string())?)
        .knots(to_knots(&spec.knots))
        .interp(parse_interp(spec.interp.as_deref())?);
    if let Some(day_count) = spec.day_count.as_deref() {
        builder = builder.day_count(parse_day_count(day_count)?);
    }
    builder
        .build()
        .map_err(|err| format!("build discount curve '{}': {err}", spec.id))
}

fn build_forward_curve(spec: &ForwardCurveSpec) -> Result<ForwardCurve, String> {
    let mut builder = ForwardCurve::builder(spec.id.as_str(), spec.tenor)
        .base_date(parse_as_of_date(&spec.base_date).map_err(|err| err.to_string())?)
        .knots(to_knots(&spec.knots))
        .interp(parse_interp(spec.interp.as_deref())?);
    if let Some(day_count) = spec.day_count.as_deref() {
        builder = builder.day_count(parse_day_count(day_count)?);
    }
    builder
        .build()
        .map_err(|err| format!("build forward curve '{}': {err}", spec.id))
}

fn build_inflation_curve(spec: &InflationCurveSpec) -> Result<InflationCurve, String> {
    let mut builder = InflationCurve::builder(spec.id.as_str())
        .base_date(parse_as_of_date(&spec.base_date).map_err(|err| err.to_string())?)
        .base_cpi(spec.base_cpi)
        .knots(to_knots(&spec.knots))
        .interp(parse_interp(spec.interp.as_deref())?);
    if let Some(day_count) = spec.day_count.as_deref() {
        builder = builder.day_count(parse_day_count(day_count)?);
    }
    if let Some(lag) = spec.indexation_lag_months {
        builder = builder.indexation_lag_months(lag);
    }
    builder
        .build()
        .map_err(|err| format!("build inflation curve '{}': {err}", spec.id))
}

fn build_hazard_curve(spec: &HazardCurveSpec) -> Result<HazardCurve, String> {
    let mut builder = HazardCurve::builder(spec.id.as_str())
        .base_date(parse_as_of_date(&spec.base_date).map_err(|err| err.to_string())?)
        .knots(to_knots(&spec.knots));
    if let Some(recovery_rate) = spec.recovery_rate {
        builder = builder.recovery_rate(recovery_rate);
    }
    if let Some(day_count) = spec.day_count.as_deref() {
        builder = builder.day_count(parse_day_count(day_count)?);
    }
    builder
        .build()
        .map_err(|err| format!("build hazard curve '{}': {err}", spec.id))
}

fn to_knots(knots: &[[f64; 2]]) -> Vec<(f64, f64)> {
    knots.iter().map(|knot| (knot[0], knot[1])).collect()
}

fn parse_day_count(raw: &str) -> Result<DayCount, String> {
    DayCount::from_str(raw).map_err(|err| format!("invalid day_count '{raw}': {err}"))
}

fn parse_interp(raw: Option<&str>) -> Result<InterpStyle, String> {
    raw.unwrap_or("linear")
        .parse::<InterpStyle>()
        .map_err(|err| format!("invalid curve interpolation: {err}"))
}
